Source/Core/VideoBackends/Software/TextureSampler.cpp

```diff
@@ -18,29 +18,32 @@
 
 namespace TextureSampler
 {
-static inline void WrapCoord(int* coordp, WrapMode wrapMode, int imageSize)
+static inline void WrapCoord(int* coordp, WrapMode wrap_mode, int image_size)
 {
   int coord = *coordp;
-  switch (wrapMode)
+  switch (wrap_mode)
   {
   case WrapMode::Clamp:
-    coord = (coord > imageSize) ? imageSize : (coord < 0) ? 0 : coord;
+    coord = std::clamp(coord, 0, image_size - 1);
     break;
   case WrapMode::Repeat:
-    coord = coord % (imageSize + 1);
-    coord = (coord < 0) ? imageSize + coord : coord;
+    // Per YAGCD's info on TX_SETMODE1_I0 (et al.), mirror "requires the texture size to be a power
+    // of two. (wrapping is implemented by a logical AND (SIZE-1))".  So though this doesn't wrap
+    // nicely for non-power-of-2 sizes, that's how hardware does it.
+    coord = coord & (image_size - 1);
+    coord = (coord < 0) ? image_size - 1 + coord : coord;
```

Comment:
The `(coord < 0)` case is no longer needed, as the bitwise and will always result in a positive value.

I suspect that this is also responsible for all of the rendering differences I mentioned above (and initially incorrectly attributed to size + 1), which should also bring the software renderer in line with the hardware renderers.